File: python/basics/intervals/57_insert_interval.py
'''
https://leetcode.com/problems/insert-interval/description/?envType=study-plan-v2&envId=top-interview-150

https://www.youtube.com/watch?v=A8NUOmlwOlM
#interesting

iterate the original list linearly,
    3 cases
        1) the new_interval ends before the interval => add the new node.. add the rest
        2) the new_interval starts after the end of interval => add the interval
        3) the new interval overlaps with the interval, create the new_inteval : min of starts, max of ends
            dont add the new interval.  Add it in the end of the loop
'''


def insert_interval_linear(intervals: list[list[int]], new_interval:list[int]):
    result = []

    for i in range(len(intervals)):
        #new ends before next start - non overlapping
        if new_interval[1] < intervals[i][0]:
            result.append(new_interval)
            return result + intervals[i:]
        #non-overlapping - starts before the next one
        elif new_interval[0] > intervals[i][1]:#nonoverlapping
            result.append(intervals[i])
        else:
            new_interval = [min(new_interval[0], intervals[i][0]),
                            max(new_interval[1], intervals[i][1])]
    result.append(new_interval)
    return result

def insert_interval(intervals: list[list[int]], new_interval:list[int]):
    len_intervals = len(intervals)
    l = 0
    r = len_intervals-1
    result = intervals.copy()
    #binary search to find the insertion point
    index = -1
    while l <= r:
        m = l + (r-l)//2
        if new_interval[0] == intervals[m][0]:
            index = m
            break
        elif new_interval[0] > intervals[m][0]:
            l = m+1
        else:
            r = m-1
    if index == -1:
        index = r
    #print("index: " + str(intervals[index]) + str(intervals[l]) + str(intervals[m]) )

    #result = intervals.copy()
    #result.insert(l, new_interval)
    result = []

    for i in range(len(intervals)):
        if not result or (result[-1][1] < new_interval[0]):
            result.append(i)
        else:
            result[-1][1] = max(result[-1][1], new_interval[1])
    return result

intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]]
new_interval = [4,8]
#Output: [[1,2],[3,10],[12,16]]
#Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
print("intervals-before: " + str(intervals))
result = insert_interval_linear(intervals, new_interval)
print("intervals-after: " + str(result))


intervals = [[1,3],[6,9]]
new_interval = [2,5]
#Output: [[1,5],[6,9]]
#Example 2:
print("intervals-before: " + str(intervals))
result = insert_interval_linear(intervals, new_interval)
print("intervals-after: " + str(result))

File: problem_solving/intervals/cpp/57_insert_interval.cpp
/*
https://leetcode.com/problems/insert-interval/
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"


std::vector<std::vector<int>> insert_interval(const std::vector<std::vector<int>>& intervals,
                                                std::vector<int>& newInterval ) {
    std::vector<std::vector<int>> results;
    int len = intervals.size();
    bool inserted = false;
    for(int i = 0; i < len; ++i) {
        if(newInterval[1] < intervals[i][0]) {
            if(!inserted) {
                results.push_back(newInterval);
                inserted = true;
            }
            results.push_back(intervals[i]);
        } else if(newInterval[0] > intervals[i][1]) {
            results.push_back(intervals[i]);
        } else {
            newInterval[0] = std::min(newInterval[0], intervals[i][0]);
            newInterval[1] = std::max(newInterval[1], intervals[i][1]);
        }
    }
    if(!inserted) {
        results.push_back(newInterval);
    }
    return results;
}


int main() {
    std::vector<std::vector<int>> intervals;
    intervals = {{1,2},{3,5},{6,7},{8,10},{12,16}};
    std::vector<int> new_interval = {4,8};
    //#Output: [[1,2],[3,10],[12,16]]
    //#Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
    //print_vector_of_vector(intervals);
    std::vector<std::vector<int>> result = insert_interval(intervals, new_interval);
    print_vector_of_vector(result);


    intervals = {{1,3},{6,9}};
    new_interval = {2,5};
    //#Output: [[1,5],[6,9]]
    //#Example 2:
    print_vector_of_vector(intervals);
    result = insert_interval(intervals, new_interval);
    print_vector_of_vector(result);
    return 0;
}